NuRadioReco/utilities/noise.py

```diff
@@ -508,15 +553,22 @@ def generate_noise(self, phasing_mode="slice", trigger_mode="binned_sum", debug=
             dt_triggering += time.process_time() - tstart
 
             if is_triggered:
-                triggered_bin = triggered_bin // 2  # the trace is cut in the downsampled version. Therefore, triggered bin is factor of two smaller. 
+                triggered_bin = triggered_bin // self.upsampling  # the trace is cut in the downsampled version. Therefore, triggered bin is factor of two smaller. 
                 i_low = triggered_bin - self.pre_trigger_bins
                 i_high = i_low + self.n_samples_trigger
-                if (i_low >= 0) and (i_high < self.n_samples):
-                    # traces need to be downsampled
-                    # resample and use axis -1 since trace might be either shape (N) for analytic trace or shape (3,N) for E-field
-                    self._traces = scipy.signal.resample(self._traces, np.shape(self._traces)[-1] // self.upsampling, axis=-1)
+
+                # traces need to be downsampled
+                # resample and use axis -1 since trace might be either shape (N) for analytic trace or shape (3,N) for E-field
+                self._traces = scipy.signal.resample(self._traces, np.shape(self._traces)[-1] // self.upsampling, axis=-1)
+
+                if (i_low >= 0) and (i_high < self.n_samples): # If range is directly a subset of the waveform
                     return self._traces[:, i_low:i_high], self._phased_traces
 
+                # Otherwise, roll the waveforms. Safe as long as noise is generated in the freq domain
+                self._phased_traces = np.roll(self._phased_traces, -i_low * self.upsampling, axis=-1)
+                self._traces = np.roll(self._traces, -i_low, axis=-1)
+                return self._traces[:, :self.n_samples_trigger], self._phased_traces, triggered_beam
```

Comment:
These two `return` statements have different signatures (and neither corresponds to the one specified in the docstring, which should probably be updated?)
Re rolling - the idea is that the generated `self._traces` is periodic because the generation is in the frequency domain, and the resampling happens after the triggering step & doesn't introduce artifacts at the edges?

Reply:
Fixed the signature.

This is correct. If you generated, band-limited noise, then a roll is equivalent to a Fourier shift. The magnitude of the frequency content remains unchanged. So it still has the same band, phase-delays, etc